tests/scripts/thread-cert/Cert_5_2_04_REEDUpgrade.py

```diff
@@ -82,14 +99,79 @@ def test(self):
             time.sleep(5)
             self.assertEqual(self.nodes[i].get_state(), 'router')
 
-        self.nodes[REED].start()
+        # 2 DUT_REED: Attach to ROUTER, 2-hops from the Leader.
+        self.nodes[DUT_REED].start()
         time.sleep(5)
-        self.assertEqual(self.nodes[REED].get_state(), 'child')
+        time.sleep(ROUTER_SELECTION_JITTER)
+
+        reed_messages = self.sniffer.get_messages_sent_by(DUT_REED)
+
+        # The DUT_REED must not send a coap message here.
+        msg = reed_messages.does_not_contain_coap_message()
+        assert msg is True, "Error: The REED sent an Address Solicit Request"
+
+        # 3 DUT_REED: Check if DUT_REED sent a correct MLE Advertisement or not.
+        msg = reed_messages.next_mle_message(mle.CommandType.ADVERTISEMENT)
+        msg.assertSentWithHopLimit(255)
+        msg.assertSentToDestinationAddress('ff02::1')
+        msg.assertMleMessageContainsTlv(mle.SourceAddress)
+        msg.assertMleMessageContainsTlv(mle.LeaderData)
+        msg.assertMleMessageDoesNotContainTlv(mle.Route64)
+
+        # 4 Wait for DUT_REED to send the second package.
+        time.sleep(REED_ADVERTISEMENT_INTERVAL + REED_ADVERTISEMENT_MAX_JITTER)
 
+        # 5 DUT_REED: Check if it sent the second MLE Advertisement.
+        reed_messages = self.sniffer.get_messages_sent_by(DUT_REED)
+        msg = reed_messages.next_mle_message(mle.CommandType.ADVERTISEMENT)
+
+        # 6 ED
         self.nodes[ED].start()
         time.sleep(5)
-        self.assertEqual(self.nodes[ED].get_state(), 'child')
-        self.assertEqual(self.nodes[REED].get_state(), 'router')
+
+        # 7 DUT_REED: Check if it sent a correct MLE Parent Response.
+        reed_messages = self.sniffer.get_messages_sent_by(DUT_REED)
+        msg = reed_messages.next_mle_message(mle.CommandType.PARENT_RESPONSE)
+        msg.assertSentToNode(self.nodes[ED])
+        msg.assertMleMessageContainsTlv(mle.SourceAddress)
+        msg.assertMleMessageContainsTlv(mle.LeaderData)
+        msg.assertMleMessageContainsTlv(mle.LinkLayerFrameCounter)
+        msg.assertMleMessageContainsOptionalTlv(mle.MleFrameCounter)
+        msg.assertMleMessageContainsTlv(mle.Response)
+        msg.assertMleMessageContainsTlv(mle.Challenge)
+        msg.assertMleMessageContainsTlv(mle.LinkMargin)
+        msg.assertMleMessageContainsTlv(mle.Connectivity)
+        msg.assertMleMessageContainsTlv(mle.Version)
+
+        # 8 DUT_REED: Verify that its Address Solicit Request is properly.
```

Comment:
suggestion: you can simplify to "8 DUT_REED: Verify Address Solicit Request"

You can simplify the other comments in the same way.

Reply:
fixed:)